oe_contact_forms.module

```diff
@@ -6,3 +6,350 @@
+ * @param array $form
+ *   The form.
+ * @param Drupal\Core\Form\FormStateInterface $form_state
+ *   The form state.
+ *
+ * @return array
+ *   Topics fieldset.
+ */
+function _oe_contact_forms_add_more_callback(array &$form, FormStateInterface $form_state): array {
+  return $form['corporate_fields']['topics_fieldset'];
+}
+
+/**
+ * Add element to the form based on ajax.
+ *
+ * @param array $form
+ *   The form.
+ * @param Drupal\Core\Form\FormStateInterface $form_state
+ *   The form state.
+ */
+function _oe_contact_forms_add_one_callback(array &$form, FormStateInterface $form_state): void {
+  $num_topics = $form_state->get('num_topics');
+  $num_topics++;
+  $form_state->set('num_topics', $num_topics);
+  $form_state->setRebuild();
+}
+
+/**
+ * Remove element from the form based on ajax.
+ *
+ * @param array $form
+ *   The form.
+ * @param Drupal\Core\Form\FormStateInterface $form_state
+ *   The form state.
+ */
+function _oe_contact_forms_remove_callback(array &$form, FormStateInterface $form_state): void {
+  $num_topics = $form_state->get('num_topics');
+
+  if ($num_topics > 1) {
+    $num_topics--;
+  }
+
+  $form_state->set('num_topics', $num_topics);
+  $form_state->setRebuild();
+}
+
+/**
+ * Ajax callback: Return markup for 'corporate_fields'.
+ *
+ * According 'is_corporate_form' value.
+ *
+ * @param array $form
+ *   The form.
+ * @param Drupal\Core\Form\FormStateInterface $form_state
+ *   The form state.
+ *
+ * @return array
+ *   Corporate fields.
+ */
+function _oe_contact_forms_is_corporate_form_ajax_callback(array &$form, FormStateInterface $form_state): array {
+  return $form['corporate_fields'];
+}
+
+/**
+ * Entity builder for the contact form edit form with third party options.
+ *
+ * @param string $entity_type
+ *   The contact form entity.
+ * @param \Drupal\contact\ContactFormInterface $contact_form
+ *   The contact form entity.
+ * @param array $form
+ *   The form.
+ * @param Drupal\Core\Form\FormStateInterface $form_state
+ *   The form state.
+ */
+function _oe_contact_forms_contact_form_builder($entity_type, ContactFormInterface $contact_form, array &$form, FormStateInterface $form_state): void {
+  if ((boolean) $form_state->getValue('is_corporate_form')) {
+    // Prevent duplicated topic values on "Add topic" ajax.
+    $contact_form->setThirdPartySetting('oe_contact_forms', 'topics', []);
+  }
+
+  // Prevent invalid arguments in Drupal\contact\ContactFormEditForm::form()L96.
+  $recipients = $contact_form->getRecipients();
+
+  if (gettype($recipients) !== 'array') {
+    $contact_form->setRecipients((array) $recipients);
+  }
+}
+
+/**
+ * Submit callback for 'contact_form' entity: Manage custom fields data.
+ *
+ * @param array $form
+ *   The form.
+ * @param Drupal\Core\Form\FormStateInterface $form_state
+ *   The form state.
+ *
+ * @throws \Drupal\Core\Entity\EntityStorageException
+ */
+function _oe_contact_forms_contact_form_submit(array &$form, FormStateInterface $form_state): void {
```

Comment:
Since we are on an entity form which uses an entity builder for both submit and validation, you can move this entire logic to the entity builder no? And then in the validate you can access the values directly in the third party settings.

Also, it removes the need for saving the config twice in the same operation -> causes two config save events, etc..
Would this work?

Reply:
lets discuss